libdrgn/program.c

```diff
@@ -667,32 +749,98 @@ struct drgn_error *drgn_program_cache_prstatus_entry(struct drgn_program *prog,
 		entry->str = data;
 		entry->len = size;
 	} else {
-		struct drgn_prstatus_map_entry entry = {
-			.value = { data, size },
-		};
-		struct drgn_error *err = get_prstatus_pid(prog, data, size,
-							  &entry.key);
+		struct drgn_thread thread = { .prog = prog, .prstatus = { data, size } };
+		struct drgn_error *err = get_prstatus_pid(prog, data, size, &thread.tid);
 		if (err)
 			return err;
-		if (drgn_prstatus_map_insert(&prog->prstatus_map, &entry,
-					     NULL) == -1)
+		*ret = thread.tid;
+		if (drgn_thread_set_insert(&prog->thread_set, &thread, NULL) == -1)
 			return &drgn_enomem;
 	}
 	return NULL;
 }
 
-static struct drgn_error *drgn_program_cache_prstatus(struct drgn_program *prog)
+static struct drgn_error *drgn_program_cache_threads(struct drgn_program *prog);
+
+static struct drgn_error *
+drgn_program_kernel_coredump_cache_crashed_thread(struct drgn_program *prog)
+{
+	assert((prog->flags & DRGN_PROGRAM_IS_LINUX_KERNEL) &&
+	       !(prog->flags & DRGN_PROGRAM_IS_LIVE));
+	assert(prog->threads_cached);
+	if (prog->crashed_thread)
+		return NULL;
+	struct drgn_error *err;
+	uint32_t crashed_thread_tid;
+	int64_t crashing_cpu_value;
+	struct drgn_object crashing_cpu;
+	drgn_object_init(&crashing_cpu, prog);
+	err = drgn_program_find_object(prog, "crashing_cpu", NULL,
+				       DRGN_FIND_OBJECT_VARIABLE,
+				       &crashing_cpu);
+	if (err)
+		goto out;
+	err = drgn_object_read_signed(&crashing_cpu, &crashing_cpu_value);
+	if (err)
+		goto out;
+	if (crashing_cpu_value < 0 ||
+	    crashing_cpu_value >= prog->prstatus_vector.size) {
+		err = drgn_error_create(
+			DRGN_ERROR_LOOKUP,
+			"crashed thread not found, debug info is possibly corrupted");
+		goto out;
+	}
+	struct nstring *prstatus =
+		&prog->prstatus_vector.data[crashing_cpu_value];
+	err = get_prstatus_pid(prog, prstatus->str, prstatus->len,
+			       &crashed_thread_tid);
+	if (err)
+		goto out;
+	if (crashed_thread_tid == 0) {
+		prog->crashed_thread = malloc(sizeof(*prog->crashed_thread));
+		if (!prog->crashed_thread) {
+			err = &drgn_enomem;
+			goto out;
+		}
+		prog->crashed_thread->prog = prog;
+		prog->crashed_thread->tid = crashed_thread_tid;
+		drgn_object_init(&prog->crashed_thread->object, prog);
+		err = linux_helper_idle_task(&prog->crashed_thread->object, crashing_cpu_value);
+		if (err) {
+			drgn_object_deinit(&prog->crashed_thread->object);
+			free(prog->crashed_thread);
```

Comment:
`prog->crashed_thread` needs to be reset to `NULL` here (either by moving labels around or by setting it explicitly).